### ISSUE TITLE
How to evaluate AMI?

### ISSUE THREAD

**Issue by xp1632** [xp1632/Aalen_working_log#65]
### Strategy
 
- We peeled off the [pdf extraction module] and [Chaldene visualization module] in evaluation as they are unimportant for performance metrics
- And only keep the core of this method for evalution, LLM+Gradient Descent

**Comment by xp1632** [comment]
### Contribution of this work

- To think about what to evaluate, we need to think about the `selling point` of the paper

- Title idea: `LLM-Guided Differentiable Visual Programming for Image Processing`

- Or: `Hybrid LLM-Gradient Optimization for Interpretable Image Enhancement`
---

- Our selling point is `novel hybrid system`: unique combination of

- LLM generates the workflow structure, gradient descent optimizes the parameters, visual programming provides interpretability

---

**- Key Claims:**

  - LLM provides intelligent workflow structure selection
  - Gradient descent provides precise parameter optimization
  - Visual programming provides interpretability and user control
  - Works zero-shot with single before/after example
---

PDF (before/after images + task description)
    ↓
┌─────────────────────────────────────────┐
│  Step 1: LLM (GPT-4)                    │
│  - Analyzes task description            │
│  - Generates initial workflow structure │
│  - Selects operations (blur, contrast)  │
└─────────────────────────────────────────┘
    ↓
┌─────────────────────────────────────────┐
│  Step 2: Gradient Optimization          │
│  - Takes LLM's workflow                 │
│  - Fine-tunes exact parameter values    │
│  - Uses before/after as loss signal     │
└─────────────────────────────────────────┘
    ↓
┌─────────────────────────────────────────┐
│  Step 3: Chaldene Visual Workflow       │
│  - Converts to visual representation    │
│  - Real-time animation of optimization  │
│  - Interpretable, editable by user      │
└─────────────────────────────────────────┘

---


Component | GenArtist | VisProg | InstructPix2Pix | Ours
-- | -- | -- | -- | --
LLM selects operations | ✅ | ✅ | ❌ | ✅
Gradient optimizes params | ❌ | ❌ | ❌ | ✅
Visual workflow output | ❌ | ❌ | ❌ | ✅
Single example (zero-shot) | ❌ | ❌ | ❌ | ✅
Interpretable parameters | ❌ | Partial | ❌ | ✅

**Comment by xp1632** [comment]
### Goal of Evaluation

**- Goal:** Show that LLM + Gradient is better than either alone.

------
## Phase 1: Create 10 Synthetic Test Cases with Known Ground Truth

- [x] synthetic test
- [ ] record occurred issues
- [ ] fix quick ones

 


 - Use  10 images. Apply known transforms:



| Test ID | Operations | Ground Truth Params |
|---------|------------|---------------------|
| T01 | DiffBrightness | factor=0.7 |
| T02 | DiffContrast | factor=1.5 |
| T03 | GaussianBlur2D | radius=5.0 |
| T04 | DiffBrightness + DiffContrast | 0.8, 1.3 |
| T05 | GaussianBlur2D + DiffBrightness | 3.0, 1.2 |
| T06 | DiffGamma | gamma=1.5 |
| T07 | UnsharpMask | radius=4.0 |
| T08 | DiffBrightness + DiffContrast + DiffGamma | 0.9, 1.2, 1.3 |
| T09 | GaussianBlur2D + UnsharpMask | 4.0, 2.0 |
| T10 | Full chain (4 ops) | various |

<img width="1616" height="561" alt="Image" src="https://github.com/user-attachments/assets/6b789591-28e6-486b-8b23-4182988d29d0" />

**Comment by xp1632** [comment]
## Phase 2: Three Methods to Compare

 

### Method A: LLM Only (No Gradient)

 

```

1. Give LLM the task description

2. LLM outputs workflow with guessed parameters

3. DON'T run gradient optimization

4. Measure: How close are LLM's guesses to ground truth?

```

 

### Method B: Gradient Only (Manual Workflow)

 

```

1. Manually specify the correct workflow structure

2. Run gradient optimization

3. Measure: Can gradient find the right parameters?

```

 

### Method C: Your System (LLM + Gradient)

 

```

1. LLM generates workflow

2. Gradient optimizes parameters

3. Measure: Best of both?

```

 

---

 

## Phase 3: What to Measure

 

### For Each Test Case, Record:

| Metric | How to Compute |
|--------|----------------|
| **PSNR** | `skimage.metrics.peak_signal_noise_ratio(target, output)` |
| **SSIM** | `skimage.metrics.structural_similarity(target, output)` |
| **Param Error** | `abs(predicted - ground_truth)` for each param |
| **Time** | Seconds to complete |

**Comment by xp1632** [comment]
## Phase 4: Expected Results

 

If your system works, you should see:

 

| Method | Param Error | PSNR | Notes |
|--------|-------------|------|-------|
| LLM Only | High (~30%) | Low | LLM guesses are imprecise |
| Gradient Only | Low (<5%) | High | But needs manual workflow |
| **LLM + Gradient** | Low (<5%) | High | Best: auto workflow + precision |

 

**Key insight to show:**

- LLM alone: Wrong parameters, bad output

- Your system: LLM's structure + gradient's precision = good output

**Comment by xp1632** [comment]
### V1 Evaluation and Issues (Only MSE gradient descent)

<img width="1117" height="601" alt="Image" src="https://github.com/user-attachments/assets/da5d24eb-12b4-4a73-af5c-e2a72f630900" />

<img width="732" height="433" alt="Image" src="https://github.com/user-attachments/assets/88136a6d-a81d-4c25-9781-a65315c86ade" />

<img width="878" height="662" alt="Image" src="https://github.com/user-attachments/assets/af549dc4-2468-4802-ade5-7bc5c72e7a69" />

-----

### 
- The NaN output in T05 caused by small gamma in DiffGamma is fixed by:

<img width="1327" height="996" alt="Image" src="https://github.com/user-attachments/assets/c5a18f1a-7401-4b5c-a02f-a9d956d51d24" />

**Comment by xp1632** [comment]
### V2 Evaluation (LLM generated workflow + MSE gradient descent optimized parameter)

### Preliminary conclusion:

1. LLM+GD >  only LLM 
2. In certain cases, LLM+GD > GD, because LLM suggests one more Op(extra flexibility), so we have one more parameter to optimize
3. In certain cases, LLM+ GD behaves badly because of bad Op choice in LLM 

"Our LLM+Gradient system achieves 48.5 dB average PSNR, surpassing even the oracle baseline (46.5 dB) by discovering richer workflows. The LLM suggests complementary operations that, when optimized together, achieve better results than manually specified pipelines. However, destructive operations (e.g., blur) can limit optimization potential, suggesting future work in operation selection."

<img width="1351" height="360" alt="Image" src="https://github.com/user-attachments/assets/e112dd9a-27d2-4b1f-ae73-4f79b5576aee" />

---
### Result 

```
Found 10 test cases: ['T01', 'T02', 'T03', 'T04', 'T05', 'T06', 'T07', 'T08', 'T09', 'T10']
================================================================================

T01:
  Baseline:        PSNR=15.98
  GT workflow: [('DiffBrightness', [-1], {'factor': [1.0]})]
  Gradient-only:   PSNR=45.84
  Task description: I need to make the image darker (reduce brightness).
  LLM workflow: [('DiffBrightness', [-1], {'factor': [0.5]}), ('DiffContrast', [-1], {'factor': [0.8]})]
  LLM-only:        PSNR=18.11
  LLM workflow: [('DiffBrightness', [-1], {'factor': [0.5]}), ('DiffContrast', [-1], {'factor': [0.8]})]
  LLM+Gradient:    PSNR=49.70

T02:
  Baseline:        PSNR=17.87
  GT workflow: [('DiffContrast', [-1], {'factor': [1.0]})]
  Gradient-only:   PSNR=43.38
  Task description: I need to increase contrast.
  LLM workflow: [('DiffContrast', [-1], {'factor': [1.5]}), ('DiffBrightness', [-1], {'factor': [1.1]})]
  LLM-only:        PSNR=26.32
  LLM workflow: [('DiffContrast', [-1], {'factor': [1.5]}), ('DiffBrightness', [-1], {'factor': [1.1]})]
  LLM+Gradient:    PSNR=47.45

T03:
  Baseline:        PSNR=17.54
  GT workflow: [('DiffGamma', [-1], {'gamma': [1.0]})]
  Gradient-only:   PSNR=43.20
  Task description: I need to apply gamma correction to darken midtones.
  LLM workflow: [('DiffGamma', [-1], {'gamma': [1.5]}), ('DiffContrast', [-1], {'factor': [1.1]})]
  LLM-only:        PSNR=35.61
  LLM workflow: [('DiffGamma', [-1], {'gamma': [1.5]}), ('DiffContrast', [-1], {'factor': [1.1]})]
  LLM+Gradient:    PSNR=47.41

T04:
  Baseline:        PSNR=18.40
  GT workflow: [('DiffBrightness', [-1], {'factor': [1.0]}), ('DiffContrast', [-1], {'factor': [1.0]})]
  Gradient-only:   PSNR=46.37
  Task description: I need to make the image darker (reduce brightness), and increase contrast.
  LLM workflow: [('DiffBrightness', [-1], {'factor': [0.8]}), ('DiffContrast', [-1], {'factor': [1.5]})]
  LLM-only:        PSNR=32.19
  LLM workflow: [('DiffBrightness', [-1], {'factor': [0.8]}), ('DiffContrast', [-1], {'factor': [1.5]})]
  LLM+Gradient:    PSNR=50.71

T05:
  Baseline:        PSNR=18.29
  GT workflow: [('DiffBrightness', [-1], {'factor': [1.0]}), ('DiffGamma', [-1], {'gamma': [1.0]})]
  Gradient-only:   PSNR=48.09
  Task description: I need to make the image brighter (increase brightness), and apply gamma correction to brighten midtones.
  LLM workflow: [('DiffBrightness', [-1], {'factor': [1.2]}), ('DiffGamma', [-1], {'gamma': [0.8]})]
  LLM-only:        PSNR=52.82
  LLM workflow: [('DiffBrightness', [-1], {'factor': [1.2]}), ('DiffGamma', [-1], {'gamma': [0.8]})]
  LLM+Gradient:    PSNR=54.54

T06:
  Baseline:        PSNR=20.66
  GT workflow: [('DiffContrast', [-1], {'factor': [1.0]})]
  Gradient-only:   PSNR=52.60
  Task description: I need to reduce contrast.
  LLM workflow: [('DiffContrast', [-1], {'factor': [0.5]}), ('GaussianBlur2D', [-1], {'radius': [2.0]})]
  LLM-only:        PSNR=23.92
  LLM workflow: [('DiffContrast', [-1], {'factor': [0.5]}), ('GaussianBlur2D', [-1], {'radius': [2.0]})]
  LLM+Gradient:    PSNR=36.57

T07:
  Baseline:        PSNR=13.87
  GT workflow: [('DiffGamma', [-1], {'gamma': [1.0]})]
  Gradient-only:   PSNR=54.54
  Task description: I need to apply gamma correction to darken midtones.
  LLM workflow: [('DiffGamma', [-1], {'gamma': [1.5]}), ('DiffContrast', [-1], {'factor': [1.1]})]
  LLM-only:        PSNR=21.32
  LLM workflow: [('DiffGamma', [-1], {'gamma': [1.5]}), ('DiffBrightness', [-1], {'factor': [0.8]})]
  LLM+Gradient:    PSNR=49.18

T08:
  Baseline:        PSNR=19.55
  GT workflow: [('DiffBrightness', [-1], {'factor': [1.0]}), ('DiffContrast', [-1], {'factor': [1.0]}), ('DiffGamma', [-1], {'gamma': [1.0]})]
  Gradient-only:   PSNR=42.78
  Task description: I need to make the image darker (reduce brightness), and increase contrast, and apply gamma correction to darken midtones.
  LLM workflow: [('DiffBrightness', [-1], {'factor': [0.8]}), ('DiffContrast', [-1], {'factor': [1.5]}), ('DiffGamma', [-1], {'gamma': [1.8]})]
  LLM-only:        PSNR=21.09
  LLM workflow: [('DiffBrightness', [-1], {'factor': [0.8]}), ('DiffContrast', [-1], {'factor': [1.5]}), ('DiffGamma', [-1], {'gamma': [1.8]})]
  LLM+Gradient:    PSNR=46.53

T09:
  Baseline:        PSNR=16.56
  GT workflow: [('DiffBrightness', [-1], {'factor': [1.0]})]
  Gradient-only:   PSNR=46.59
  Task description: I need to make the image brighter (increase brightness).
  LLM workflow: [('DiffBrightness', [-1], {'factor': [1.5]}), ('DiffContrast', [-1], {'factor': [1.1]})]
  LLM-only:        PSNR=25.47
  LLM workflow: [('DiffBrightness', [-1], {'factor': [1.5]}), ('DiffContrast', [-1], {'factor': [1.1]})]
  LLM+Gradient:    PSNR=50.08

T10:
  Baseline:        PSNR=22.03
  GT workflow: [('DiffContrast', [-1], {'factor': [1.0]}), ('DiffGamma', [-1], {'gamma': [1.0]})]
  Gradient-only:   PSNR=41.69
  Task description: I need to increase contrast, and apply gamma correction to brighten midtones.
  LLM workflow: [('DiffContrast', [-1], {'factor': [1.5]}), ('DiffGamma', [-1], {'gamma': [0.8]})]
  LLM-only:        PSNR=30.41
  LLM workflow: [('DiffContrast', [-1], {'factor': [1.5]}), ('DiffGamma', [-1], {'gamma': [0.8]})]
  LLM+Gradient:    PSNR=53.20

================================================================================
RESULTS SUMMARY
================================================================================
Test     Baseline     LLM-only     Grad-only    LLM+Grad    
------------------------------------------------------------
T01      15.98        18.11        45.84        49.70       
T02      17.87        26.32        43.38        47.45       
T03      17.54        35.61        43.20        47.41       
T04      18.40        32.19        46.37        50.71       
T05      18.29        52.82        48.09        54.54       
T06      20.66        23.92        52.60        36.57       
T07      13.87        21.32        54.54        49.18       
T08      19.55        21.09        42.78        46.53       
T09      16.56        25.47        46.59        50.08       
T10      22.03        30.41        41.69        53.20       
------------------------------------------------------------
AVG      18.08        28.73        46.51        48.54       
================================================================================
```

---

### Analysis

<img width="1384" height="467" alt="Image" src="https://github.com/user-attachments/assets/2bb0d1d4-0906-49ec-8070-6fa1badba751" />

<img width="1406" height="532" alt="Image" src="https://github.com/user-attachments/assets/2f58f1ee-68db-413f-aeea-aa6f9f17f44d" />

<img width="1367" height="320" alt="Image" src="https://github.com/user-attachments/assets/0bc89a02-16c0-471e-b489-db8f734b0c88" />

<img width="1353" height="402" alt="Image" src="https://github.com/user-attachments/assets/8a162bc0-9cee-4c72-bf03-31747a7d5d49" />

---

### Rest issues to consider

- [ ] 1. We only feed LLM text prompts when letting it generate the workflow, also let LLM see images?

- [ ] 2. We only tested the Diff Ops in Kornia in evaluation, more Op possibility?

- [ ] 3. We only used MSE in gradient descent ,  complex loss functions generated by LLM?

- [ ] 4. We only used PNSR as final metric, add SSIM as well?

**Comment by xp1632** [comment]
### Add NN baseline

https://github.com/xp1632/Aalen_working_log/issues/69
